src/cargo/sources/git/known_hosts.rs

```diff
@@ -316,42 +386,81 @@ fn check_ssh_known_hosts_loaded(
-        // This indicates the key has changed.
-        // This is only reported as an error if no subsequent lines have a
-        // correct key.
-        changed_key = Some(known_host.clone());
     }
-    // Older versions of OpenSSH (before 6.8, March 2015) showed MD5
-    // fingerprints (see FingerprintHash ssh config option). Here we only
-    // support SHA256.
-    let mut remote_fingerprint = cargo_util::Sha256::new();
-    remote_fingerprint.update(remote_host_key);
-    let remote_fingerprint =
-        base64::encode_config(remote_fingerprint.finish(), base64::STANDARD_NO_PAD);
-    let remote_host_key = base64::encode(remote_host_key);
-    // FIXME: Ideally the error message should include the IP address of the
-    // remote host (to help the user validate that they are connecting to the
-    // host they were expecting to). However, I don't see a way to obtain that
-    // information from libgit2.
-    match changed_key {
-        Some(old_known_host) => Err(KnownHostError::HostKeyHasChanged {
+
+    if latent_errors.len() == 0 {
+        // FIXME: Ideally the error message should include the IP address of the
+        // remote host (to help the user validate that they are connecting to the
+        // host they were expecting to). However, I don't see a way to obtain that
+        // information from libgit2.
+        Err(KnownHostError::HostKeyNotFound {
             hostname: host.to_string(),
             key_type: remote_key_type,
-            old_known_host,
-            remote_host_key,
-            remote_fingerprint,
-        }),
-        None => Err(KnownHostError::HostKeyNotFound {
-            hostname: host.to_string(),
-            key_type: remote_key_type,
-            remote_host_key,
+            remote_host_key: remote_host_key_encoded,
             remote_fingerprint,
             other_hosts,
-        }),
+        })
+    } else {
+        // We're going to take the first HostKeyHasChanged error if
+        // we find one, otherwise we'll take the first error (which
+        // we expect to be a CertAuthority error).
+        for err in &latent_errors {
+            if let KnownHostError::HostKeyHasChanged {
+                hostname,
+                key_type,
+                old_known_host,
+                remote_host_key,
+                remote_fingerprint,
+            } = err
+            {
+                return Err(KnownHostError::HostKeyHasChanged {
+                    hostname: hostname.clone(),
+                    key_type: key_type.clone(),
+                    old_known_host: old_known_host.clone(),
+                    remote_host_key: remote_host_key.clone(),
+                    remote_fingerprint: remote_fingerprint.clone(),
+                });
+            }
+        }
+        // Otherwise, we take the first error (which we expect to be
+        // a CertAuthority error).
+        Err(latent_errors.pop().unwrap())
```

Comment:
Just to cut down on the number of lines of code here, I suggest just removing the matching element instead of trying to clone it:

```suggestion
        if let Some(index) = latent_errors
            .iter()
            .position(|e| matches!(e, KnownHostError::HostKeyHasChanged { .. }))
        {
            return Err(latent_errors.remove(index));
        } else {
            // Otherwise, we take the first error (which we expect to be
            // a CertAuthority error).
            Err(latent_errors.pop().unwrap())
        }
```

Reply:
Thank you! I struggled with this bit and wasn't happy with the solution, although I couldn't come up with anything better myself. This is much nicer!